### DISCUSSION TITLE
Put _deactivate_missing_dags logic behind config

### DISCUSSION THREAD

**Discussion by adamd01** [apache/airflow#21368] [General] [answered]
### Description

In Airflow 2.2.0 this PR was added (https://github.com/apache/airflow/pull/17121) which deactivates any dags which are no longer found in the file being processed. 

It would be good to have the option to make this logic configurable.

### Use case/motivation

We have a set up where we generate DAGs dynamically by hitting an API. (In order to abstract DAG creation from users who can't directly code up an Airflow DAG).

We have 100s of DAGs which means importing all the DAGs each time the dag processor runs is slow. And we want to make sure any new DAGs are added as quickly as possible. 

To make this happens we use some state on the scheduler to record when the last dag parse happened, and then hit our api with that timestamp to only add new dags. Every 5 minutes we do a full refresh to make sure everything is in sync. 

With this change in place we have to go back to doing the full import each time, which means we have a slow process for importing new dags. Right now this stops us from upgrading to 2.2.x

As a potential workaround I tried syncing serialised DAGs from the DB to supplement the incremental API fetches, but this seems quite messy and threw up some other issues. 

I'd be keen to know if anyone has other ideas of a work around or if there is more support for this functionality to be configurable I'm happy to work on a PR. 

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
I turned it into discussion because this won't be handled as a "feature" I am afraid.

I think we will not apply such change to implement the specific case of yours? If you want to propose a bigger change in processing of Airflow DAG - discussion on devlist and AIP (maybe even follow-up on https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-5+Remote+DAG+Fetcher) 

How about using modified time? https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#file-parsing-sort-mode ? Did it work for you? I think you can also follow https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html?highlight=fine%20tuning#fine-tuning-your-scheduler-performance - to see if you can improve the perfomance of scheduler.

**Reply by potiuk** [reply]
And BTW. I do not really understand why "_deactivate_missing_dags" is a problem - are you impossibole to keep the DAG history clean ? What is your real problem to follow the "standard" pattern of airflow where dags are added and old dags are remaining where they were?

**Reply by adamd01** [reply]
Thanks for the response

In my case sorting the dag files in a different way (like modified time) won't help as all the dags are generated from a single file. 

In terms of following the standard pattern.- because we are generating so many dags in a single file by pulling definitions from the API it's just a bit slow, which I why I want to avoid generating all the dags on each parsing loop 

The remote DAG fetcher AIP sounds really good, although I'm not sure it will help in this case since all DAGs are generated from a single file. (In the file we hit an API, retrieve a custom JSON representation of what we need to generate a dag, then we generate an actual DAG object) that pattern helps us with code re-use and standardisation of implementation (as well as enabling users who aren't engineers to automate simple workflows)

It feels like this is quite a big change in the way DAG parsing works and I'd be surprised that this won't impact others

**Reply by potiuk** [reply]
In your case you could actually implement a bit different approach. It seems that you "Know" when your files changed and you could store - the cache of those parsed DAGs in S3 or smth (exported via pickle or dill library for example) and instead of building the dags every time parsing happens, you could restore those dags from the last cache if you are sure they will stay the same - rather than re-building them.

This way you coudl avoid costly operations in most cases where you (on your own) determine, there is no point in rebuilding the DAG. 

This is mich easier for you - than for Airflow parser to do, because you know where your DAG came from and what are the "variables" for it and you could likely determine an easy way to decide when to invalidate the cache. Actually this is the main reason why we DO reparse the DAGs every time - because in Airlfow core we actually have no way to reliably determine if the cache should be invalidated - but you can do it.

As the old saying  is

"There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors."  - this is the 1st problem in its full swing.

**Reply by adamd01** [reply]
Good idea - I started attempting to do this by using Airflow's DB as that cache - since serialised dags are stored there in json. 
I ran across some issues with it, which I've not fully ironed out yet, but I guess I'll go back to that option.

As you pointed out the way I can use this to invalidate them out of that cache is useful and I had been looking for a mechanism to more easily remove DAGs, so I suppose all in all this probably is better for me too in the end. 

For anyone who comes across this thread, what I'm attempting is:


```
def get_serialised_dags_from_db(filename: str) -> {str: DAG}:
    """ Get serialised dags out of the Airflow DB
    :param filename: The name of the .py file which generated the DAGs
    :return: dictionary of {dag_id: DAG}
    """
    dag_bag = DagBag(read_dags_from_db=True)
    dag_bag.collect_dags_from_db()
    dags = {}
    for dag_id, dag in dag_bag.dags.items():
        if filename in dag.filepath:
            if not dag.edge_info:
                dag.edge_info = {}  # fixes an issue where this isn't always set
            dags[dag_id] = dag
    return dags
```

Thanks for the help @potiuk 👍

**Reply by potiuk** [reply]
> Good idea - I started attempting to do this by using Airflow's DB as that cache - since serialised dags are stored there in json.
> I ran across some issues with it, which I've not fully ironed out yet, but I guess I'll go back to that option.

Just a problem with that - the serialized DAGs in the Airflow DB are NOT ENOUGH in this case. The serialized form in the DB is enough to keep the structre of the DAG, parameters, configuration etc - but it's not nearly enough to re-create DAG Python Object (which is what you need at this stage. The Python -> Serialized is one-way only and not reversible. What you really need to do is to serialize Python Object to a binary form and restore it back. Which is what `dill` library might help you with.

**Reply by potiuk** [reply]
But on the other hand, I guess it might be a nice feature in this case to somehow mark  a DAG as "no-need to re-parse"  during parsing. Maybe an interesting feature to consider - Maybe you could experiment with this a bit and that actually MIGHT be an interesting optimization feature we could consider adding.

**Reply by adamd01** [reply]
ah right, that probably explains why it isn't quite working for me yet. Thanks, that should save me some time going down the wrong path!

**Reply by adamd01** [reply]
I like your suggestion, that would be neat
I'll have a play about with that

**Comment by adamd01** [comment, marked as answer]
Update on this, I went with @potiuk suggestion of implementing a local dag cache using dill and it's working nicely
We now get the benefit that dags deleted upstream are also deleted in Airflow, and no performance downsides

In case it's of use, here's an idea of the code we're running:

```
import dill as pickle
from configparser import ConfigParser

CONFIG_PATH = "dag_generator.ini"
CONFIG_SECTION = 'dag_generator'


def set_persistent_variable(name: str, value: str):
    """ Use config parser to store persisted variables on the box """
    if value is None:  # can't set None, need to remove
        return unset_persistent_variable(name)
    config = ConfigParser()
    config.read(CONFIG_PATH)
    if not config.has_section(CONFIG_SECTION):
        config.add_section(CONFIG_SECTION)
    config.set(CONFIG_SECTION, name, value)
    with open(CONFIG_PATH, 'w') as fp:
        config.write(fp)


def unset_persistent_variable(name: str):
    """ Delete from the config store """
    try:
        config = ConfigParser()
        config.read(CONFIG_PATH)
        if config.has_section(CONFIG_SECTION):
            config.remove_option(CONFIG_SECTION, name)
        with open(CONFIG_PATH, 'w') as fp:
            config.write(fp)
    except Exception:
        # Value probably doesn't exist
        pass


def get_perisistent_variable(name: str, default=None):
    """ Use config parser to store persited variables on the box """
    try:
        config = ConfigParser()
        config.read(CONFIG_PATH)
        return config.get(CONFIG_SECTION, name)
    except Exception:
        # probably not set yet
        return default


def cache_dags(dags: {str: DAG}):
    """ Pickle dags to a file, to be retreived later """
    flush_cached_dags()
    with tempfile.NamedTemporaryFile(prefix=DAG_CACHE_PREFIX, delete=False) as dag_cache:
        logging.info("saving dags to cache '%s'" % dag_cache.name)
        set_persistent_variable(DAG_CACHE_LOCATION, dag_cache.name)
        pickle.dump(dags, dag_cache)


def flush_cached_dags():
    """ Drop/invalidate the whole cache """
    dag_cache_location = get_perisistent_variable(DAG_CACHE_LOCATION)
    if dag_cache_location and os.path.isfile(dag_cache_location):
        os.remove(dag_cache_location)
    set_persistent_variable(DAG_CACHE_LOCATION, None)


def get_cached_dags() -> {str: DAG}:
    """ Load cached dags from pickle store """
    dag_cache_location = get_perisistent_variable(DAG_CACHE_LOCATION)
    if dag_cache_location and os.path.exists(dag_cache_location):
        with open(dag_cache_location, 'rb') as dag_cache:
            print("Getting cached dags from '%s'" % dag_cache_location)
            return pickle.load(dag_cache)
    else:
        logging.warning(f"Couldn't find any cached dags")
        return {}
```

And then in the loop where I get dags from an API we do something like this:

```
 def run() -> {DAG}:
    """ Runs the loop for getting and building dags """
    full_refresh, ts, last_updated = is_full_refresh()

    if full_refresh:
        flush_cached_dags()  # This allows old dags to be deleted
        dags = {}
    else:
        dags = get_cached_dags()

    dags_json = get_dags_from_api(full_refresh, last_updated)

    for dag_json in dags_json:
        try:
            _dag = generate_DAG(dag_json)
            dags[dag_definition.get('dag_id')] = _dag
            sync_perm_for_dag(_dag, users)
        except Exception as e:
            logging.exception(f"Dag failed to load {dag_json}")
            if os.getenv(AIRFLOW_IS_K8S_EXECUTOR_POD):
                # load in a dummy that can show the error to the user
                _dag = fallback_dummy_dag(exception=e)
                logging.info(f"Dummy dag built: {_dag}")
                dags[_dag.dag_id] = _dag

    record_refresh_status(full_refresh, ts)
    cache_dags(dags)
    return dags
```

**Reply by potiuk** [reply]
Nice!

Also we actually have another change coming in Airflow 2.3: https://github.com/apache/airflow/pull/21399 - this one might also help with handling the problem. Rather than deactivating missing dags immediately, it delays a bit both - scanning for those DAGs as well as add config to see how long time they are missing and only deactivates them when they are not parsed for whole interval. 

So in case you have short periods when the DAGs are not available at least. Not 100% sure if it handles the problem of yours, the local dag cache is nice though :)
